File: src/main/java/universalelectricity/compatibility/ModuleBuildCraft.java
package universalelectricity.compatibility;

import net.minecraft.item.ItemStack;
import net.minecraftforge.common.ForgeDirection;
import universalelectricity.api.CompatibilityModule;
import universalelectricity.api.CompatibilityType;
import buildcraft.api.power.IPowerReceptor;
import buildcraft.api.power.PowerHandler.PowerReceiver;
import buildcraft.api.power.PowerHandler.Type;

/**
 * @author Calclavia
 * 
 */
public class ModuleBuildCraft extends CompatibilityModule
{
	@Override
	public long doReceiveEnergy(Object handler, ForgeDirection direction, long energy, boolean doReceive)
	{
		IPowerReceptor receptor = ((IPowerReceptor) handler);
		PowerReceiver receiver = receptor.getPowerReceiver(direction);

		if (receiver != null)
		{
			if (doReceive)
			{
				return (long) (receiver.receiveEnergy(Type.PIPE, (float) (energy * CompatibilityType.BUILDCRAFT.ratio), direction) * CompatibilityType.BUILDCRAFT.reciprocal_ratio);
			}

			return (long) (receiver.powerRequest() * CompatibilityType.BUILDCRAFT.reciprocal_ratio);
		}

		return 0;
	}

	@Override
	public long doExtractEnergy(Object handler, ForgeDirection direction, long energy, boolean doExtract)
	{
		return 0;
	}

	@Override
	public boolean doIsHandler(Object obj)
	{
		return obj instanceof IPowerReceptor;
	}

	@Override
	public boolean doCanConnect(Object obj, ForgeDirection direction, Object source)
	{
		return ((IPowerReceptor) obj).getPowerReceiver(direction) != null;
	}

	@Override
	public long doChargeItem(ItemStack itemStack, long joules, boolean docharge)
	{
		return 0;
	}

	@Override
	public long doDischargeItem(ItemStack itemStack, long joules, boolean doDischarge)
	{
		return 0;
	}

	@Override
	public ItemStack doGetItemWithCharge(ItemStack itemStack, long energy)
	{
		return null;
	}

	@Override
	public boolean doIsEnergyContainer(Object obj)
	{
		return obj instanceof IPowerReceptor;
	}

	@Override
	public long doGetEnergy(Object obj, ForgeDirection direction)
	{
		if (obj instanceof IPowerReceptor)
		{
			if (((IPowerReceptor) obj).getPowerReceiver(direction) != null)
			{
				return (long) (((IPowerReceptor) obj).getPowerReceiver(direction).getEnergyStored() * CompatibilityType.BUILDCRAFT.reciprocal_ratio);
			}
		}

		return 0;
	}

	@Override
	public long doGetMaxEnergy(Object obj, ForgeDirection direction)
	{
		if (obj instanceof IPowerReceptor)
		{
			if (((IPowerReceptor) obj).getPowerReceiver(direction) != null)
			{
				return (long) (((IPowerReceptor) obj).getPowerReceiver(direction).getMaxEnergyStored() * CompatibilityType.BUILDCRAFT.reciprocal_ratio);
			}
		}

		return 0;
	}

	@Override
	public long doGetEnergyItem(ItemStack is)
	{
		return 0;
	}

	@Override
	public long doGetMaxEnergyItem(ItemStack is)
	{
		return 0;
	}
}

File: src/main/scala/universalelectricity/compatibility/module/ModuleBuildCraft.scala
package universalelectricity.compatibility.module

import buildcraft.api.mj.{IBatteryProvider}
import net.minecraft.item.ItemStack
import net.minecraftforge.common.util.ForgeDirection
import universalelectricity.compatibility.Compatibility

/**
 * Module for BuildCraft support
 * @author Calclavia
 */
object ModuleBuildCraft extends Compatibility.CompatibilityModule("BuildCraft|Energy", "BuildCraft", "Minecraft Joule", "MJ", 500)
{
  def receiveEnergy(handler: AnyRef, direction: ForgeDirection, energy: Double, doReceive: Boolean): Double =
  {
    val receiver = handler.asInstanceOf[IBatteryProvider].getMjBattery("buildcraft.kinesis")

    if (receiver != null)
    {
      if (doReceive)
      {
        return receiver.addEnergy(energy * ratio) * reciprocal_ratio
      }

      return (receiver.getEnergyRequested * reciprocal_ratio)
    }
    return 0
  }

  def extractEnergy(handler: AnyRef, direction: ForgeDirection, energy: Double, doExtract: Boolean): Double =
  {
    val receiver = handler.asInstanceOf[IBatteryProvider].getMjBattery("buildcraft.kinesis")

    if (receiver != null)
    {
      if (doExtract)
      {
        return receiver.addEnergy(energy * ratio) * reciprocal_ratio
      }

      return energy
    }

    return 0
  }

  def doIsHandler(obj: AnyRef): Boolean =
  {
    return obj.isInstanceOf[IBatteryProvider]
  }

  def canConnect(obj: AnyRef, direction: ForgeDirection, source: AnyRef): Boolean =
  {
    return obj.asInstanceOf[IBatteryProvider].getMjBattery("buildcraft.kinesis") != null
  }

  def chargeItem(itemStack: ItemStack, joules: Double, docharge: Boolean): Double =
  {
    return 0
  }

  def dischargeItem(itemStack: ItemStack, joules: Double, doDischarge: Boolean): Double =
  {
    return 0
  }

  def getItemWithCharge(itemStack: ItemStack, energy: Double): ItemStack =
  {
    return null
  }

  def doIsEnergyContainer(obj: AnyRef): Boolean =
  {
    return obj.isInstanceOf[IBatteryProvider]
  }

  def getEnergy(obj: AnyRef, direction: ForgeDirection): Double =
  {
    val receiver = obj.asInstanceOf[IBatteryProvider].getMjBattery("buildcraft.kinesis")

    if (receiver != null)
    {
      return receiver.getEnergyStored * reciprocal_ratio
    }

    return 0
  }

  def getMaxEnergy(obj: AnyRef, direction: ForgeDirection): Double =
  {
    val receiver = obj.asInstanceOf[IBatteryProvider].getMjBattery("buildcraft.kinesis")

    if (receiver != null)
    {
      return receiver.maxCapacity * reciprocal_ratio
    }
    return 0
  }

  def getEnergyItem(is: ItemStack): Double =
  {
    return 0
  }

  def getMaxEnergyItem(is: ItemStack): Double =
  {
    return 0
  }
}